Feature/spacy detector

aCampello opened the pull request:
Current stats for the old entity detector:

```bash
Documents generated in 62.92s
Scrubbed documents in  3.98s
                           precision    recall  f1-score   support

      name           name       0.09      1.00      0.16       558
```

Current stats for new spacy detector
```bash
Documents generated in 66.99s
Scrubbed documents in  78.61s
                               precision    recall  f1-score   support

named_entity     name       0.98      0.95      0.97      1143

                    micro avg       0.98      0.95      0.97      1143
                    macro avg       0.98      0.95      0.97      1143
                 weighted avg       0.98      0.95      0.97      1143
```

thomasbird commented:
Nice work, thanks! I haven't had a chance to look through it properly, but two things popped out:
* It would be great to have some tests for the new behaviour around iter_filth_documents and the new detector. 
* I'd like to keep Filth types related to what they are as opposed how they were detected. So if we find a name a NameFilth should be returned or if we find a organisation an OrganisationFilth could be returned. The filth_cls attribute is no longer needed on the Detector class to make it possible to return any Filth you'd like. 

I look forward to playing with this next week! Thanks for the commits! 😁

aCampello commented:
> * It would be great to have some tests for the new behaviour around iter_filth_documents and the new detector.

100%. This is highly unfinished. I am planning to writing tests shortly, to get higher than it was before this PR. Just wanted to write a draft PR to get early feedback.

> * I'd like to keep Filth types related to what they are as opposed how they were detected. So if we find a name a NameFilth should be returned or if we find a organisation an OrganisationFilth could be returned. The filth_cls attribute is no longer needed on the Detector class to make it possible to return any Filth you'd like.

Yes, that is also something that occurred to me, as I was playing with it. Will implement it 👍 !

aCampello commented:
All right, I implemented the above change about returning Known filth types. Now we have a dict that maps Spacy labels to known filth classes. 

Still most of spacy NER's labels do not correspond to a filth type, so at the moment I kept `NamedEntity` filth as a 'default' filth class in these cases (you can pass a label that will further specify with Named entity it is). Would like to hear your thoughts on whether you think this is a sensible approach, or would it be better for us to: create new filth class for some of the labels (if so, which ones in https://nightly.spacy.io/models/en#en_core_web_lg-labels), or just accept `PERSON, ORG` for the time being.

Current usage:

```
In [12]: d = NamedEntityDetector(named_entities=['PERSON', 'ORG', 'GPE'])

In [13]: list(d.iter_filth("Antonio was hired by Google, but was fired by Apple. He now lives in Seattle."))
None
Out[13]: 
[<NameFilth text='Antonio' detector_name='named_entity'>,
 <OrganizationFilth text='Google' detector_name='named_entity'>,
 <OrganizationFilth text='apple Apple' detector_name='named_entity'>,
 <NamedEntityFilth text='Seattle' detector_name='named_entity'>]
```

thomasbird commented on scrubadub/detectors/named_entity.py:
```diff
@@ -0,0 +1,60 @@
+                                    detector_name=self.name,
+                                    label=ent.label_)
+
+    def iter_filth_documents(self, documents: Union[Sequence[str], Dict[str, str]]) -> Generator[Filth, None, None]:
```
Since this function is only boilerplate and would have to be repeated in each of the detectors, we could make the interface more similar to the `_iter_spacy_pipeline` function and keep this part in the scrubber, perhaps?

This would mean that the signature for iter_filth_documents would look like this:

```def iter_filth_documents(self, doc_names: Sequence[Optional[str]], doc_list: Sequence[str]) -> Generator[Filth, None, None]:```

Does that sound like a good idea?

thomasbird commented on scrubadub/scrubbers.py:
```diff
@@ -222,7 +222,8 @@ def _post_process_filth_list(self, filth_list: Sequence[Filth]) -> Sequence[Filt
     def iter_filth(
-            self, text: str, document_name: Optional[str] = None, run_post_processors: bool = True
+            self, text: str, document_name: Optional[str] = None, run_post_processors: bool = True,
+            run_merge: bool = True, exclude_detectors: Optional[List[str]] = None
```
The scrubber is looking a little complex now. What if we could simplify it a little? How about making `iter_filth` only call `iter_filth_documents` and we do all the heavy lifting in there?

thomasbird commented on scrubadub/scrubbers.py:
```diff
@@ -263,33 +271,55 @@ def iter_filth_documents(
+            document_iterator = getattr(detector, 'iter_filth_documents', None)
+            if callable(document_iterator):
+                document_detectors_names.append(name)
+                for filth in document_iterator(documents):
```
Then we can re-use most of the stuff that was in iter_filth in here like this:

```
    def iter_filth(
            self, text: str, document_name: Optional[str] = None, run_post_processors: bool = True,
    ) -> Generator[Filth, None, None]:
        """Iterate over the different types of filth that can exist.
        """
        for filth in self.iter_filth_documents(
            documents={document_name: text}, run_post_processors=run_post_processors
        ):
            yield filth

    def iter_filth_documents(
            self,
            documents: Union[Sequence[str], Dict[Optional[str], str]],
            run_post_processors: bool = True
    ) -> Generator[Filth, None, None]:
        """Iterate over the different types of filth that can exist."""
        if not isinstance(documents, (dict, list)):
            raise TypeError('documents must be one of a string, list of strings or dict of strings.')

        if isinstance(documents, dict):
            document_names, document_texts = zip(*documents.items())
        elif isinstance(documents, (tuple, list)):
            document_texts = documents
            document_names = [str(x) for x in range(len(documents))]

        # currently doing this by aggregating all_filths and then sorting
        # inline instead of with a Filth.__cmp__ method, which is apparently
        # much slower http://stackoverflow.com/a/988728/564709
        #
        # NOTE: we could probably do this in a more efficient way by iterating
        # over all detectors simultaneously. just trying to get something
        # working right now and we can worry about efficiency later
        filth_list = []  # type: List[Filth]
        for name, detector in self._detectors.items():
            if callable(getattr(detector, 'iter_filth_documents', None)):
                for filth in detector.iter_filth_documents(document_names, document_texts):
                    if not isinstance(filth, Filth):
                        raise TypeError('iter_filth must always yield Filth')
                    filth_list.append(filth)
            else:
                for document_name, text in zip(document_names, document_texts):
                    for filth in detector.iter_filth(text, document_name=document_name):
                        if not isinstance(filth, Filth):
                            raise TypeError('iter_filth must always yield Filth')
                        filth_list.append(filth)

        # This is split up so that we only have to use lists if we have to post_process Filth
        if run_post_processors:
            all_filths = list(self._merge_filths(filth_list))
            all_filths = list(self._post_process_filth_list(all_filths))

            # Here we loop over a list of Filth...
            for filth in all_filths:
                yield filth
        else:
            # ... but here, we're using a generator. If we try to use the same variable it would have two types and
            # fail static typing in mypy
            for filth in self._merge_filths(filth_list):
                yield filth
```

The above doesn't pass mypy yet, but the test cases ran. Does this successfully simplify things a bit?

aCampello replied:
That sounds good. I actually started implementing something along these lines, but since I didn't want to to mess around with the scrubber code that much, I decided to change my mind and just code "around" it. But your suggestion seems much better for organisation. I will give it a shot!

thomasbird commented:
This is looking really great! The comments above are only icing on the cake :cake: Thanks!

aCampello commented on scrubadub/detectors/named_entity.py:
```diff
@@ -0,0 +1,60 @@
+        # Spacy NER are all upper cased
+        self.named_entities = {entity.upper() for entity in named_entities}
+        if model not in spacy.info()['pipelines']:
+            raise OSError("Can't find model '{}'. If it is a valid Spacy model, "
```
This was addressed in https://github.com/LeapBeyond/scrubadub/pull/64/commits/6446f44b0543b70fa9de0413287890ef0c856d8b

thomasbird replied:
Awesome!

aCampello commented:
I have now implemented those changes pinpointed before. Also added some initial tests to the detector.

At the moment the build fails for python3.5, because Spacy doesn't support it. So wondering how we want to go about that. Should I add the spacy detector as an "extra"/should I allow the tests to fail ?

thomasbird commented:
> At the moment the build fails for python3.5, because Spacy doesn't support it
Ah. Yeah including it as an extra seems very sensible.

python 3.5 is almost end of life, but i believe it's the python still in ubuntu LTS 18.04... Don't worry about it in the MR, I'll work out something...

aCampello commented:
Hi @thomasbird , I believe I am happy with the initial functionality. I added the tests. On python 3.8 I am over 98%, however since the functionality doesn't work in 3.5, there is a decrease in coverage (I set the tests to skip).

Let me know what you think would be the best approach for coveralls. Happy to hear your hours on the MR in general.

thomasbird reviewed (approved):
Awesome work! Thanks for the commits!

aCampello commented:
Thanks for merging this. I believe this closes https://github.com/LeapBeyond/scrubadub/issues/18
